pokt-network/pocket
# [Main][Telemetry]: Merge the first part of the telemetry (metrics) milestone into main

## Description


This PR adds event-metrics to Pre2P RainTree implementation and the HotPOKT consensus implementation using the Telemetry module.

## Issues
This PR ticks off the deliverables 1, 2, 3, 4 from #101 

## Changes
Please mark the options that are relevant.

- [x] Added instrumentation to record the reception of a RainTree message as an event for each block, this event has the following attributes:
     - namespace `pre2p`
     - event_name `raintree_messages`
     - labels: [height]

- [x] Added instrumentation to record the reception of redundant RainTree messages as an event for each new block, the event has the following attributes:
     - namespace `pre2p`
     - event_name `broadcast_message_redundancy_per_block`
     - labels: [height, message_hash] 

- [x] Added `logql` query to process the recorded events and fed the results to a numerical series graph


- [x] Added instrumentation to record the reception of a hotpokt message as an event for each block, this event has the following attributes:
     - namespace `consensus`
     - event_name `hotpokt_messages`
     - labels: [height, step, validator_role]

- [x] Added logql query to process the recorded events and fed the results to a numerical series graph

## How Has This Been Tested?
__

1. Spin up the stack
```zsh
$ make compose_and_watch
```

2. Make sure you spin the client up, and that kick the blockchain off:
```zsh
$ make client_start && make client_connect
```

3. Choose:
```
✔ TogglePacemakerMode
✔ TriggerNextView
```

4. Open Grafana at [http://localhost:3000](http://localhost:3000)
5. Open the `RainTree/HotPokt Stats Dashboard v0.1`

6. Verify the results on the dashboard

### Additional items for review

The query that was used to generate the graph is the following:
```logql
sum by (blockheight) (count_over_time({host=\"docker-desktop\"}\n    |= \"[EVENT] event_metrics_namespace_consensus\"\n    | pattern `<datetime> <_> <time> <type> <_> <metric_name> <_> <blockheight> <step> <validator_role>`[6s]))"
```
6. Verify the results on the dashboard

### Additional items for review

The query that was used to generate the graph is the following:

1. Total # HotPOKT Msgs / Block
```sql
sum by (blockheight) (
   count_over_time({host="docker-desktop"}
   |= "[EVENT] consensus"
   | pattern `<datetime> <_> <time> <type> <_> <metric_name> <_> <blockheight> <step> <validator_role>`[6s]
   )
),
```

2. Avg # Raintree Msgs / Block

```sql
avg by (blockheight, instance) (
  sum by (blockheight) (
    count_over_time({host="docker-desktop"}
    |= "[EVENT] p2p raintree_messages"
    | pattern `<datetime> <_> <time> <type> <namespace> <metric> <_> <blockheight>`[6s])
  )
)
```

```sql
avg by (blockheight, instance) (
  sum by (blockheight) (
    count_over_time({host="docker-desktop"}
    |= "[EVENT] p2p broadcast_message_redundancy_per_block"
    | pattern `<datetime> <_> <time> <type> <_> <redundancy> <_> <hashstring> <_> <blockheight>`[6s])
   )
)
```


Owners: @derrandz 
Co-owners: @okdas @Olshansk
